File: sdk/javascript/src/types/privacy.ts
/**
 * Privacy level enum
 */
export enum PrivacyLevel {
  PUBLIC = 0,
  PSEUDONYMOUS = 1,
  PRIVATE = 2,
  ANONYMOUS = 3,
}

/**
 * Privacy settings
 */
export interface PrivacySettings {
  address: string;
  defaultLevel: PrivacyLevel;
  transactionPrivacy: boolean;
  balancePrivacy: boolean;
  identityPrivacy: boolean;
  mixingEnabled: boolean;
  ringSignatureSize: number;
  stealthAddressEnabled: boolean;
}

/**
 * Private transaction parameters
 */
export interface PrivateTransactionParams {
  sender: string;
  recipient: string;
  amount: string;
  denom: string;
  privacyLevel: PrivacyLevel;
  memo?: string;
}

/**
 * Confidential transaction
 */
export interface ConfidentialTransaction {
  txHash: string;
  commitment: string;
  rangeProof: string;
  sender: string; // Stealth address
  recipient: string; // Stealth address
  timestamp: Date;
  confirmations: number;
}

/**
 * Ring signature parameters
 */
export interface RingSignatureParams {
  message: string;
  signers: string[]; // Public keys
  realSignerIndex: number;
  privateKey: string;
}

/**
 * Ring signature
 */
export interface RingSignature {
  signature: string;
  ringMembers: string[];
  keyImage: string;
  timestamp: Date;
}

/**
 * Mixing configuration
 */
export interface MixingConfig {
  enabled: boolean;
  minMixSize: number;
  maxMixSize: number;
  mixingRounds: number;
  delay: number;
  fee: string;
}

/**
 * Stealth address
 */
export interface StealthAddress {
  address: string;
  publicViewKey: string;
  publicSpendKey: string;
  createdAt: Date;
  used: boolean;
}

/**
 * Privacy parameters
 */
export interface PrivacyParams {
  privacyEnabled: boolean;
  defaultPrivacyLevel: PrivacyLevel;
  ringSignatureEnabled: boolean;
  confidentialTxEnabled: boolean;
  mixingEnabled: boolean;
  stealthAddressEnabled: boolean;
  minRingSize: number;
  maxRingSize: number;
}

/**
 * Zero-knowledge proof
 */
export interface ZeroKnowledgeProof {
  proof: string;
  publicInputs: string[];
  verificationKey: string;
  proofSystem: string;
  createdAt: Date;
}

/**
 * Privacy audit log
 */
export interface PrivacyAuditLog {
  address: string;
  action: string;
  privacyLevel: PrivacyLevel;
  timestamp: Date;
  metadata: Record<string, any>;
}

File: sdk/python/aura/types/privacy.py
"""Type definitions for Privacy module."""

from dataclasses import dataclass
from typing import Optional, List, Dict
from datetime import datetime
from enum import Enum


class PrivacyLevel(Enum):
    """Privacy protection levels."""

    NONE = "none"
    BASIC = "basic"
    STANDARD = "standard"
    MAXIMUM = "maximum"


class MixingStrategy(Enum):
    """Token mixing strategies."""

    TORNADO = "tornado"
    COINJOIN = "coinjoin"
    RING_SIGNATURE = "ring_signature"
    STEALTH = "stealth"


@dataclass
class PrivacyParams:
    """Privacy operation parameters."""

    level: PrivacyLevel
    use_mixing: bool = True
    use_stealth_address: bool = False
    mixing_rounds: int = 3


@dataclass
class ConfidentialTransaction:
    """Confidential transaction details."""

    tx_id: str
    sender_commitment: str
    receiver_commitment: str
    amount_commitment: str
    range_proof: str
    created_at: datetime
    privacy_level: PrivacyLevel
    memo: Optional[str] = None


@dataclass
class StealthAddress:
    """Stealth address information."""

    address: str
    public_view_key: str
    public_spend_key: str
    created_at: datetime
    usage_count: int


@dataclass
class MixingPool:
    """Token mixing pool information."""

    pool_id: str
    token: str
    denomination: str
    total_deposits: int
    available_liquidity: str
    mixing_fee: str
    anonymity_set_size: int


@dataclass
class RingSignatureParams:
    """Ring signature parameters."""

    message: str
    ring_members: List[str]
    key_image: str
    ring_size: int


@dataclass
class RingSignature:
    """Ring signature data."""

    signature: str
    ring_members: List[str]
    key_image: str
    ring_size: int
    created_at: datetime
    verified: bool = False


@dataclass
class PrivacyProof:
    """Zero-knowledge privacy proof."""

    proof_type: str
    proof_data: str
    public_inputs: List[str]
    verified: bool
    created_at: datetime
    verifier_address: Optional[str] = None


@dataclass
class ConfidentialBalance:
    """Confidential balance information."""

    address: str
    balance_commitment: str
    blinding_factor: str
    encrypted_amount: str
    last_updated: datetime


@dataclass
class MixingTransaction:
    """Mixing transaction record."""

    mix_id: str
    strategy: MixingStrategy
    input_amount: str
    output_amount: str
    fee: str
    rounds: int
    anonymity_set: int
    started_at: datetime
    status: str
    completed_at: Optional[datetime] = None


@dataclass
class PrivacySettings:
    """Privacy configuration settings."""

    default_privacy_level: PrivacyLevel
    auto_mix: bool
    min_anonymity_set: int
    max_mixing_rounds: int
    stealth_addresses_enabled: bool
    ring_signature_size: int
